src/test/java/org/jenkins/plugins/lockableresources/LockStepTest.java

```diff
@@ -513,4 +515,89 @@ public void evaluate() throws Throwable {
 			}
 		});
 	}
+
+	@Issue("JENKINS-40879")
+	@Test
+	public void parallelLockRelease() throws Exception {
+		story.addStep(new Statement() {
+			@Override
+			public void evaluate() throws Throwable {
+				LockableResourcesManager.get().createResource("resource1");
+				LockableResourcesManager.get().createResource("resource2");
+				LockableResourcesManager.get().createResource("resource3");
+				LockableResourcesManager.get().createResource("resource4");
+				WorkflowJob j = story.j.jenkins.createProject(WorkflowJob.class, "j");
+				j.setDefinition(new CpsFlowDefinition(
+						"lock(resource: 'resource1') {\n" +
+								"    sleep 1\n" +
```

Comment:
I'm completely willing to believe the test could be better written - I spent a couple hours trying various things until I got a consistently reproducible failure, but it was the result of a lot of trial and error. The `sleep`s are needed because we do want the other runs to still be in the `lock`s when we take one run out of the last `lock`, but we don't want to dictate exactly when/which run goes for the last `lock` next, etc. Again, if you've got an alternate test approach, I'd love to replace this one, but this is what I was able to get working and after those two hours or so of experiments, I'm not hugely keen on experimenting more. =)